tools/transform.cpp

```diff
@@ -266,27 +266,34 @@ static void check_refinement(Errors &errs, Transform &t,
   // so order here matters
   src_state.startParsingPre();
   expr pre = t.precondition ? t.precondition->toSMT(src_state) : true;
-  pre &= src_state.getPre() && tgt_state.getPre();
+  pre &= tgt_state.getPre() && src_state.getPre()
+          && tgt_state.getPreOnQuantifiers();
+  expr srcpre_quant = src_state.getPreOnQuantifiers();
 
   auto [poison_cnstr, value_cnstr] = type.refines(a, b);
   expr pre_dom = pre && (dom_a && dom_b);
+  expr dom_a_b = dom_a && dom_b;
 
   Solver::check({
-    { axioms && preprocess(t, qvars, uvars, pre && dom_a.notImplies(dom_b)),
+    { axioms && preprocess(t, qvars, uvars, pre &&
+                           srcpre_quant.implies(dom_a.notImplies(dom_b))),
       [&](const Result &r) {
         err(r, false, "Source is more defined than target");
       }},
-    { axioms && preprocess(t, qvars, uvars, pre_dom && !poison_cnstr),
+    { axioms && preprocess(t, qvars, uvars, pre &&
+                           srcpre_quant.implies(dom_a_b && !poison_cnstr)),
       [&](const Result &r) {
         err(r, true, "Target is more poisonous than source");
       }},
-    { axioms && preprocess(t, qvars, uvars, pre_dom && !value_cnstr),
+    { axioms && preprocess(t, qvars, uvars, pre &&
+                           srcpre_quant.implies(dom_a_b && !value_cnstr)),
       [&](const Result &r) {
         err(r, true, "Value mismatch");
       }}
   });
 
-  if (check_expr(axioms && preprocess(t, qvars, uvars, move(pre))).isUnsat()) {
+  if (check_expr(axioms && preprocess(t, qvars, uvars, pre && srcpre_quant))
```

Comment:
this one doesn't look correct since in the previous formulas srcpre_quant is allowed to be false (thus it's under implication). Quantified vars need to change I guess (to be existential)?

Reply:
Fixed to make source quantifiers be existential.

Reply:
How about this? This would address the issue.
```
-  if (check_expr(axioms && preprocess(t, {}, {}, pre && srcpre_quant))
+  if (check_expr(axioms && preprocess(t, qvars, uvars,
+                                      srcpre_quant.implies(pre)))
```